runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/Outcome.kt

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries
+
+/**
+ * Represents the outcome of a repeated operation. This type is similar to a [Result] except it is a union type and has
+ * no flag for "success" since exceptional outcomes do not necessarily represent "failure".
+ * @param T The type of non-exception result (if present).
+ */
+sealed class Outcome<out T> {
+    /**
+     * The number of attempts executed in order to reach the outcome. Starts at 1.
+     */
+    abstract val attempts: Int
+
+    /**
+     * An outcome that includes a normal (i.e., non-exceptional) response.
+     * @param T The type of result.
+     * @param attempts The number of attempts executed in order to reach the outcome. Starts at 1.
+     * @param response The response to an operation.
+     */
+    data class ResponseOutcome<out T>(override val attempts: Int, val response: T) : Outcome<T>()
```

Comment:
suggestion

Since these are namespaced under the class `Outcome` they are going to be generally accessed as `Outcome.ResponseOutcome` thus we can probably just drop the `Outcome` suffix and have `Response` and `Exception` right? 

e.g.

```kotlin
val outcome = client.waitForSomeOutcome(...) 

when (outcome) {
    is Outcome.Response -> ...
    is Outcome.Error -> ...
}
```



Also should it be `Error` instead of `Exception`? The "error" terminology matches that of smithy (e.g. operation errors or waiter match `errorType`, etc). Not a big deal just came to mind while looking over the spec.

Reply:
Yes, `Outcome.Response` is much cleaner. Will change.

I chose "exception" here because that's a language-level concept that conveys a different meaning than the higher-level term "error", which to me implies divergence from correctness. For waiters that succeed on an exception, it seemed more correct to say that the _outcome_ was an _exception_ (which developers expect under certain conditions) rather than that the _outcome_ was an _error_ (which implies that it did not succeed).

Reply:
I mean `Exception` and `Error` are synonymous to me. I personally see no difference here and think that either way we do it will probably confuse someone because it is a nuanced point no matter how we represent it. `Exception` is fine though.